# Intro
I wanted to thrive in yahoo's fantasy basketball's 9-cat game.  
In the 9-cat game, I'm a "drafter" drafting basketball "players".  
As a drafter, we draft 13 players in total in a snake-draft with the other 11 drafters (12 drafters in total).   
We compete against other drafter's selected team.  
The 9 categories to compete against an opponent drafter in a 9-cat game are:  
- FG%
- FT%
- 3PTM
- Points
- Assists
- Rebounds
- Steals
- Blocks
- Turnovers
Win at least 5 categories accumlated in a whole week against an opponent drafter.   

By looking at players stats everyday costs me too much time.  
I want to design a project that can assist with my decision making in drafting players.  

This Project aims at preprocessing raw data scraped from https://www.basketball-reference.com/  
Main steps in this project: 
- Data Scraping and Preprocessing:
    - Includes EDA, filtering and joining tables using pandas.   
- Create a Team Roster Auto-Builder, based on:  
    - Weekly Player Performance Prediction.  
    - Dynamic Programming for team roster selection.  


Notes: 
Fantasy basketball drafting is actually such a complex contrained optimization problem that I would like to dig deep in another project (Multiple Knapsack Assignment Problem).  
Right now I just want to "assist" my decision making rather than creating a whole automated roster building algorithm.  



# 1. Data Scraping and Preprocessing

In [1]:
# Use basketball_reference_web_scraper package.  https://jaebradley.github.io/basketball_reference_web_scraper/  
from basketball_reference_web_scraper import client
from basketball_reference_web_scraper.data import OutputType
import pandas as pd

In [2]:
# players_season_totals
df_pst2018=pd.DataFrame.from_dict(client.players_season_totals(season_end_year=2018))
df_pst2019=pd.DataFrame.from_dict(client.players_season_totals(season_end_year=2019))
df_pst2021=pd.DataFrame.from_dict(client.players_season_totals(season_end_year=2021))
df_pst2022=pd.DataFrame.from_dict(client.players_season_totals(season_end_year=2022))
df_pst2023=pd.DataFrame.from_dict(client.players_season_totals(season_end_year=2023))
df_pst2024=pd.DataFrame.from_dict(client.players_season_totals(season_end_year=2024))

In [3]:
# view the first 5 rows of the 2018 data  
df_pst2018.head()

slug           name                  positions  age  \
0  abrinal01   Álex Abrines  [Position.SHOOTING_GUARD]   24   
1    acyqu01     Quincy Acy   [Position.POWER_FORWARD]   27   
2  adamsst01   Steven Adams          [Position.CENTER]   24   
3  adebaba01    Bam Adebayo          [Position.CENTER]   20   
4  afflaar01  Arron Afflalo  [Position.SHOOTING_GUARD]   32   

                         team  games_played  games_started  minutes_played  \
0  Team.OKLAHOMA_CITY_THUNDER            75              8            1134   
1          Team.BROOKLYN_NETS            70              8            1359   
2  Team.OKLAHOMA_CITY_THUNDER            76             76            2487   
3             Team.MIAMI_HEAT            69             19            1368   
4          Team.ORLANDO_MAGIC            53              3             682   

   made_field_goals  attempted_field_goals  ...  made_free_throws  \
0               115                    291  ...                39   
1               130                    365  ...                49   
2               448                    712  ...               160   
3               174                    340  ...               129   
4                65                    162  ...                22   

   attempted_free_throws  offensive_rebounds  defensive_rebounds  assists  \
0                     46                  26                  88       28   
1                     60                  40                 217       57   
2                    286                 384                 301       88   
3                    179                 118                 263      101   
4                     26                   4                  62       30   

   steals  blocks  turnovers  personal_fouls  points  
0      38       8         25             124     353  
1      33      29         60             149     411  
2      92      78        128             215    1056  
3      32      41         66             138     477  
4       4       9         21              56     179  

[5 rows x 22 columns]

Data cleaning and aggregation.  
The raw data only have columns made_field_goals, attempted_field_goals.  We want the aggregated result of made_field_goals/attempted_field_goals to get the Field Goal Percentage (FGperc)  
We also want to convert the total accumulated data into daily average data.  

In [5]:
df2018 = df_pst2018
df2018 = df2018.drop(columns=['positions', 'team'])
df2018 = df2018.groupby(['slug', 'name', 'age'], as_index=False)[['made_field_goals', 'attempted_field_goals', 'made_free_throws', 'attempted_free_throws', \
                         'offensive_rebounds', 'defensive_rebounds', 'games_played', 'points', 'assists', 'steals', \
                            'blocks', 'turnovers', 'made_three_point_field_goals']].sum()
df2018['FGperc'] = df2018['made_field_goals'] / df2018['attempted_field_goals']
df2018['FTperc'] = df2018['made_free_throws'] / df2018['attempted_free_throws']
df2018['rebs'] = df2018['offensive_rebounds'] + df2018['defensive_rebounds']
df2018['ppg'] = df2018['points'] / df2018['games_played']
df2018['rpg'] = df2018['rebs'] / df2018['games_played']
df2018['apg'] = df2018['assists'] / df2018['games_played']
df2018['spg'] = df2018['steals'] / df2018['games_played']
df2018['bpg'] = df2018['blocks'] / df2018['games_played']
df2018['topg'] = df2018['turnovers'] / df2018['games_played']
df2018['3pg'] = df2018['made_three_point_field_goals'] / df2018['games_played']

Let's have a brief understanding of what value of each categories count as good or bad.  
We use the quantiles of the data to determine three levels: elite/average/poor.  
We do it on the years 2018~2024.  
Note that data in year 2024 represents the current season that is update everyday.  

In [6]:
print("\n2018 (within top 120)")
print(f'FG%: elite: {df2018["FGperc"].sort_values(ascending=False).head(120).quantile(0.75).round(decimals=3)}, \
      avg: {df2018["FGperc"].sort_values(ascending=False).head(120).quantile(0.5).round(decimals=3)}, poor: {df2018["FGperc"].sort_values(ascending=False).head(120).quantile(0.25).round(decimals=3)}')
print(f'FT%: elite: {df2018["FTperc"].sort_values(ascending=False).head(120).quantile(0.75).round(decimals=3)}, \
      avg: {df2018["FTperc"].sort_values(ascending=False).head(120).quantile(0.5).round(decimals=3)}, poor: {df2018["FTperc"].sort_values(ascending=False).head(120).quantile(0.25).round(decimals=3)}')
print(f'3PTM: elite: {df2018["3pg"].sort_values(ascending=False).head(120).quantile(0.75).round(decimals=3)}, \
      avg: {df2018["3pg"].sort_values(ascending=False).head(120).quantile(0.5).round(decimals=3)}, poor: {df2018["3pg"].sort_values(ascending=False).head(120).quantile(0.25).round(decimals=3)}')
print(f'Points: elite: {df2018["ppg"].sort_values(ascending=False).head(120).quantile(0.75).round(decimals=3)}, \
      avg: {df2018["ppg"].sort_values(ascending=False).head(120).quantile(0.5).round(decimals=3)}, poor: {df2018["ppg"].sort_values(ascending=False).head(120).quantile(0.25).round(decimals=3)}')
print(f'Assists: elite: {df2018["apg"].sort_values(ascending=False).head(120).quantile(0.75).round(decimals=3)}, \
      avg: {df2018["apg"].sort_values(ascending=False).head(120).quantile(0.5).round(decimals=3)}, poor: {df2018["apg"].sort_values(ascending=False).head(120).quantile(0.25).round(decimals=3)}')
print(f'Rebounds: elite: {df2018["rpg"].sort_values(ascending=False).head(120).quantile(0.75).round(decimals=3)}, \
      avg: {df2018["rpg"].sort_values(ascending=False).head(120).quantile(0.5).round(decimals=3)}, poor: {df2018["rpg"].sort_values(ascending=False).head(120).quantile(0.25).round(decimals=3)}')
print(f'Steals: elite: {df2018["spg"].sort_values(ascending=False).head(120).quantile(0.75).round(decimals=3)}, \
      avg: {df2018["spg"].sort_values(ascending=False).head(120).quantile(0.5).round(decimals=3)}, poor: {df2018["spg"].sort_values(ascending=False).head(120).quantile(0.25).round(decimals=3)}')
print(f'Blocks: elite: {df2018["bpg"].sort_values(ascending=False).head(120).quantile(0.75).round(decimals=3)}, \
      avg: {df2018["bpg"].sort_values(ascending=False).head(120).quantile(0.5).round(decimals=3)}, poor: {df2018["bpg"].sort_values(ascending=False).head(120).quantile(0.25).round(decimals=3)}')
print(f'Turnovers: elite: {df2018["topg"].quantile(0.25).round(decimals=3)}, \
      avg: {df2018["topg"].quantile(0.5).round(decimals=3)}, poor: {df2018["topg"].quantile(0.75).round(decimals=3)}')

# accumulate (low total: blk, stl, turnover) -> total
# average (high total: points, ast, 3pg, reb) -> top 120


2018 (within top 120)
FG%: elite: 0.591,       avg: 0.56, poor: 0.524
FT%: elite: 0.904,       avg: 0.868, poor: 0.85
3PTM: elite: 2.318,       avg: 1.925, poor: 1.648
Points: elite: 19.427,       avg: 15.937, poor: 13.874
Assists: elite: 5.229,       avg: 3.993, poor: 3.191
Rebounds: elite: 7.665,       avg: 6.305, poor: 5.356
Steals: elite: 1.462,       avg: 1.149, poor: 1.028
Blocks: elite: 1.098,       avg: 0.828, poor: 0.652
Turnovers: elite: 0.485,       avg: 0.94, poor: 1.45


In [7]:
df2019 = df_pst2019
df2019 = df2019.drop(columns=['positions', 'team'])
df2019 = df2019.groupby(['slug', 'name', 'age'], as_index=False)[['made_field_goals', 'attempted_field_goals', 'made_free_throws', 'attempted_free_throws', \
                         'offensive_rebounds', 'defensive_rebounds', 'games_played', 'points', 'assists', 'steals', \
                            'blocks', 'turnovers', 'made_three_point_field_goals']].sum()
df2019['FGperc'] = df2019['made_field_goals'] / df2019['attempted_field_goals']
df2019['FTperc'] = df2019['made_free_throws'] / df2019['attempted_free_throws']
df2019['rebs'] = df2019['offensive_rebounds'] + df2019['defensive_rebounds']
df2019['ppg'] = df2019['points'] / df2019['games_played']
df2019['rpg'] = df2019['rebs'] / df2019['games_played']
df2019['apg'] = df2019['assists'] / df2019['games_played']
df2019['spg'] = df2019['steals'] / df2019['games_played']
df2019['bpg'] = df2019['blocks'] / df2019['games_played']
df2019['topg'] = df2019['turnovers'] / df2019['games_played']
df2019['3pg'] = df2019['made_three_point_field_goals'] / df2019['games_played']

print("\n2019 (within top 120)")
print(f'FG%: elite: {df2019["FGperc"].sort_values(ascending=False).head(120).quantile(0.75).round(decimals=3)}, \
      avg: {df2019["FGperc"].sort_values(ascending=False).head(120).quantile(0.5).round(decimals=3)}, poor: {df2019["FGperc"].sort_values(ascending=False).head(120).quantile(0.25).round(decimals=3)}')
print(f'FT%: elite: {df2019["FTperc"].sort_values(ascending=False).head(120).quantile(0.75).round(decimals=3)}, \
      avg: {df2019["FTperc"].sort_values(ascending=False).head(120).quantile(0.5).round(decimals=3)}, poor: {df2019["FTperc"].sort_values(ascending=False).head(120).quantile(0.25).round(decimals=3)}')
print(f'3PTM: elite: {df2019["3pg"].sort_values(ascending=False).head(120).quantile(0.75).round(decimals=3)}, \
      avg: {df2019["3pg"].sort_values(ascending=False).head(120).quantile(0.5).round(decimals=3)}, poor: {df2019["3pg"].sort_values(ascending=False).head(120).quantile(0.25).round(decimals=3)}')
print(f'Points: elite: {df2019["ppg"].sort_values(ascending=False).head(120).quantile(0.75).round(decimals=3)}, \
      avg: {df2019["ppg"].sort_values(ascending=False).head(120).quantile(0.5).round(decimals=3)}, poor: {df2019["ppg"].sort_values(ascending=False).head(120).quantile(0.25).round(decimals=3)}')
print(f'Assists: elite: {df2019["apg"].sort_values(ascending=False).head(120).quantile(0.75).round(decimals=3)}, \
      avg: {df2019["apg"].sort_values(ascending=False).head(120).quantile(0.5).round(decimals=3)}, poor: {df2019["apg"].sort_values(ascending=False).head(120).quantile(0.25).round(decimals=3)}')
print(f'Rebounds: elite: {df2019["rpg"].sort_values(ascending=False).head(120).quantile(0.75).round(decimals=3)}, \
      avg: {df2019["rpg"].sort_values(ascending=False).head(120).quantile(0.5).round(decimals=3)}, poor: {df2019["rpg"].sort_values(ascending=False).head(120).quantile(0.25).round(decimals=3)}')
print(f'Steals: elite: {df2019["spg"].sort_values(ascending=False).head(120).quantile(0.75).round(decimals=3)}, \
      avg: {df2019["spg"].sort_values(ascending=False).head(120).quantile(0.5).round(decimals=3)}, poor: {df2019["spg"].sort_values(ascending=False).head(120).quantile(0.25).round(decimals=3)}')
print(f'Blocks: elite: {df2019["bpg"].sort_values(ascending=False).head(120).quantile(0.75).round(decimals=3)}, \
      avg: {df2019["bpg"].sort_values(ascending=False).head(120).quantile(0.5).round(decimals=3)}, poor: {df2019["bpg"].sort_values(ascending=False).head(120).quantile(0.25).round(decimals=3)}')
print(f'Turnovers: elite: {df2019["topg"].quantile(0.25).round(decimals=3)}, \
      avg: {df2019["topg"].quantile(0.5).round(decimals=3)}, poor: {df2019["topg"].quantile(0.75).round(decimals=3)}')


2019 (within top 120)
FG%: elite: 0.595,       avg: 0.548, poor: 0.518
FT%: elite: 0.889,       avg: 0.864, poor: 0.841
3PTM: elite: 2.313,       avg: 1.911, poor: 1.611
Points: elite: 20.683,       avg: 16.673, poor: 14.186
Assists: elite: 5.523,       avg: 4.029, poor: 3.279
Rebounds: elite: 8.272,       avg: 6.688, poor: 5.45
Steals: elite: 1.404,       avg: 1.132, poor: 1.0
Blocks: elite: 1.11,       avg: 0.823, poor: 0.666
Turnovers: elite: 0.529,       avg: 0.856, poor: 1.452


In [8]:
df2021 = df_pst2021
df2021 = df2021.drop(columns=['positions', 'team'])
df2021 = df2021.groupby(['slug', 'name', 'age'], as_index=False)[['made_field_goals', 'attempted_field_goals', 'made_free_throws', 'attempted_free_throws', \
                         'offensive_rebounds', 'defensive_rebounds', 'games_played', 'points', 'assists', 'steals', \
                            'blocks', 'turnovers', 'made_three_point_field_goals']].sum()
df2021['FGperc'] = df2021['made_field_goals'] / df2021['attempted_field_goals']
df2021['FTperc'] = df2021['made_free_throws'] / df2021['attempted_free_throws']
df2021['rebs'] = df2021['offensive_rebounds'] + df2021['defensive_rebounds']
df2021['ppg'] = df2021['points'] / df2021['games_played']
df2021['rpg'] = df2021['rebs'] / df2021['games_played']
df2021['apg'] = df2021['assists'] / df2021['games_played']
df2021['spg'] = df2021['steals'] / df2021['games_played']
df2021['bpg'] = df2021['blocks'] / df2021['games_played']
df2021['topg'] = df2021['turnovers'] / df2021['games_played']
df2021['3pg'] = df2021['made_three_point_field_goals'] / df2021['games_played']

print("\n2021 (within top 120)")
print(f'FG%: elite: {df2021["FGperc"].sort_values(ascending=False).head(120).quantile(0.75).round(decimals=3)}, \
      avg: {df2021["FGperc"].sort_values(ascending=False).head(120).quantile(0.5).round(decimals=3)}, poor: {df2021["FGperc"].sort_values(ascending=False).head(120).quantile(0.25).round(decimals=3)}')
print(f'FT%: elite: {df2021["FTperc"].sort_values(ascending=False).head(120).quantile(0.75).round(decimals=3)}, \
      avg: {df2021["FTperc"].sort_values(ascending=False).head(120).quantile(0.5).round(decimals=3)}, poor: {df2021["FTperc"].sort_values(ascending=False).head(120).quantile(0.25).round(decimals=3)}')
print(f'3PTM: elite: {df2021["3pg"].sort_values(ascending=False).head(120).quantile(0.75).round(decimals=3)}, \
      avg: {df2021["3pg"].sort_values(ascending=False).head(120).quantile(0.5).round(decimals=3)}, poor: {df2021["3pg"].sort_values(ascending=False).head(120).quantile(0.25).round(decimals=3)}')
print(f'Points: elite: {df2021["ppg"].sort_values(ascending=False).head(120).quantile(0.75).round(decimals=3)}, \
      avg: {df2021["ppg"].sort_values(ascending=False).head(120).quantile(0.5).round(decimals=3)}, poor: {df2021["ppg"].sort_values(ascending=False).head(120).quantile(0.25).round(decimals=3)}')
print(f'Assists: elite: {df2021["apg"].sort_values(ascending=False).head(120).quantile(0.75).round(decimals=3)}, \
      avg: {df2021["apg"].sort_values(ascending=False).head(120).quantile(0.5).round(decimals=3)}, poor: {df2021["apg"].sort_values(ascending=False).head(120).quantile(0.25).round(decimals=3)}')
print(f'Rebounds: elite: {df2021["rpg"].sort_values(ascending=False).head(120).quantile(0.75).round(decimals=3)}, \
      avg: {df2021["rpg"].sort_values(ascending=False).head(120).quantile(0.5).round(decimals=3)}, poor: {df2021["rpg"].sort_values(ascending=False).head(120).quantile(0.25).round(decimals=3)}')
print(f'Steals: elite: {df2021["spg"].sort_values(ascending=False).head(120).quantile(0.75).round(decimals=3)}, \
      avg: {df2021["spg"].sort_values(ascending=False).head(120).quantile(0.5).round(decimals=3)}, poor: {df2021["spg"].sort_values(ascending=False).head(120).quantile(0.25).round(decimals=3)}')
print(f'Blocks: elite: {df2021["bpg"].sort_values(ascending=False).head(120).quantile(0.75).round(decimals=3)}, \
      avg: {df2021["bpg"].sort_values(ascending=False).head(120).quantile(0.5).round(decimals=3)}, poor: {df2021["bpg"].sort_values(ascending=False).head(120).quantile(0.25).round(decimals=3)}')
print(f'Turnovers: elite: {df2021["topg"].quantile(0.25).round(decimals=3)}, \
      avg: {df2021["topg"].quantile(0.5).round(decimals=3)}, poor: {df2021["topg"].quantile(0.75).round(decimals=3)}')


2021 (within top 120)
FG%: elite: 0.618,       avg: 0.556, poor: 0.519
FT%: elite: 0.933,       avg: 0.891, poor: 0.868
3PTM: elite: 2.736,       avg: 2.216, poor: 1.931
Points: elite: 21.933,       avg: 17.595, poor: 14.91
Assists: elite: 5.82,       avg: 4.442, poor: 3.509
Rebounds: elite: 7.932,       avg: 6.659, poor: 5.727
Steals: elite: 1.282,       avg: 1.115, poor: 1.0
Blocks: elite: 1.142,       avg: 0.932, poor: 0.749
Turnovers: elite: 0.5,       avg: 0.89, poor: 1.423


In [9]:
df2022 = df_pst2022
df2022 = df2022.drop(columns=['positions', 'team'])
df2022 = df2022.groupby(['slug', 'name', 'age'], as_index=False)[['made_field_goals', 'attempted_field_goals', 'made_free_throws', 'attempted_free_throws', \
                         'offensive_rebounds', 'defensive_rebounds', 'games_played', 'points', 'assists', 'steals', \
                            'blocks', 'turnovers', 'made_three_point_field_goals']].sum()
df2022['FGperc'] = df2022['made_field_goals'] / df2022['attempted_field_goals']
df2022['FTperc'] = df2022['made_free_throws'] / df2022['attempted_free_throws']
df2022['rebs'] = df2022['offensive_rebounds'] + df2022['defensive_rebounds']
df2022['ppg'] = df2022['points'] / df2022['games_played']
df2022['rpg'] = df2022['rebs'] / df2022['games_played']
df2022['apg'] = df2022['assists'] / df2022['games_played']
df2022['spg'] = df2022['steals'] / df2022['games_played']
df2022['bpg'] = df2022['blocks'] / df2022['games_played']
df2022['topg'] = df2022['turnovers'] / df2022['games_played']
df2022['3pg'] = df2022['made_three_point_field_goals'] / df2022['games_played']

print("\n2022 (within top 120)")
print(f'FG%: elite: {df2022["FGperc"].sort_values(ascending=False).head(120).quantile(0.75).round(decimals=3)}, \
      avg: {df2022["FGperc"].sort_values(ascending=False).head(120).quantile(0.5).round(decimals=3)}, poor: {df2022["FGperc"].sort_values(ascending=False).head(120).quantile(0.25).round(decimals=3)}')
print(f'FT%: elite: {df2022["FTperc"].sort_values(ascending=False).head(120).quantile(0.75).round(decimals=3)}, \
      avg: {df2022["FTperc"].sort_values(ascending=False).head(120).quantile(0.5).round(decimals=3)}, poor: {df2022["FTperc"].sort_values(ascending=False).head(120).quantile(0.25).round(decimals=3)}')
print(f'3PTM: elite: {df2022["3pg"].sort_values(ascending=False).head(120).quantile(0.75).round(decimals=3)}, \
      avg: {df2022["3pg"].sort_values(ascending=False).head(120).quantile(0.5).round(decimals=3)}, poor: {df2022["3pg"].sort_values(ascending=False).head(120).quantile(0.25).round(decimals=3)}')
print(f'Points: elite: {df2022["ppg"].sort_values(ascending=False).head(120).quantile(0.75).round(decimals=3)}, \
      avg: {df2022["ppg"].sort_values(ascending=False).head(120).quantile(0.5).round(decimals=3)}, poor: {df2022["ppg"].sort_values(ascending=False).head(120).quantile(0.25).round(decimals=3)}')
print(f'Assists: elite: {df2022["apg"].sort_values(ascending=False).head(120).quantile(0.75).round(decimals=3)}, \
      avg: {df2022["apg"].sort_values(ascending=False).head(120).quantile(0.5).round(decimals=3)}, poor: {df2022["apg"].sort_values(ascending=False).head(120).quantile(0.25).round(decimals=3)}')
print(f'Rebounds: elite: {df2022["rpg"].sort_values(ascending=False).head(120).quantile(0.75).round(decimals=3)}, \
      avg: {df2022["rpg"].sort_values(ascending=False).head(120).quantile(0.5).round(decimals=3)}, poor: {df2022["rpg"].sort_values(ascending=False).head(120).quantile(0.25).round(decimals=3)}')
print(f'Steals: elite: {df2022["spg"].sort_values(ascending=False).head(120).quantile(0.75).round(decimals=3)}, \
      avg: {df2022["spg"].sort_values(ascending=False).head(120).quantile(0.5).round(decimals=3)}, poor: {df2022["spg"].sort_values(ascending=False).head(120).quantile(0.25).round(decimals=3)}')
print(f'Blocks: elite: {df2022["bpg"].sort_values(ascending=False).head(120).quantile(0.75).round(decimals=3)}, \
      avg: {df2022["bpg"].sort_values(ascending=False).head(120).quantile(0.5).round(decimals=3)}, poor: {df2022["bpg"].sort_values(ascending=False).head(120).quantile(0.25).round(decimals=3)}')
print(f'Turnovers: elite: {df2022["topg"].quantile(0.25).round(decimals=3)}, \
      avg: {df2022["topg"].quantile(0.5).round(decimals=3)}, poor: {df2022["topg"].quantile(0.75).round(decimals=3)}')


2022 (within top 120)
FG%: elite: 0.644,       avg: 0.571, poor: 0.533
FT%: elite: 0.976,       avg: 0.888, poor: 0.868
3PTM: elite: 2.695,       avg: 2.23, poor: 1.906
Points: elite: 21.183,       avg: 17.301, poor: 15.015
Assists: elite: 5.767,       avg: 4.439, poor: 3.561
Rebounds: elite: 8.318,       avg: 6.399, poor: 5.497
Steals: elite: 1.341,       avg: 1.152, poor: 1.017
Blocks: elite: 1.066,       avg: 0.829, poor: 0.692
Turnovers: elite: 0.469,       avg: 0.809, poor: 1.297


In [10]:
df2023 = df_pst2023
df2023 = df2023.drop(columns=['positions', 'team'])
df2023 = df2023.groupby(['slug', 'name', 'age'], as_index=False)[['made_field_goals', 'attempted_field_goals', 'made_free_throws', 'attempted_free_throws', \
                         'offensive_rebounds', 'defensive_rebounds', 'games_played', 'points', 'assists', 'steals', \
                            'blocks', 'turnovers', 'made_three_point_field_goals']].sum()
df2023['FGperc'] = df2023['made_field_goals'] / df2023['attempted_field_goals']
df2023['FTperc'] = df2023['made_free_throws'] / df2023['attempted_free_throws']
df2023['rebs'] = df2023['offensive_rebounds'] + df2023['defensive_rebounds']
df2023['ppg'] = df2023['points'] / df2023['games_played']
df2023['rpg'] = df2023['rebs'] / df2023['games_played']
df2023['apg'] = df2023['assists'] / df2023['games_played']
df2023['spg'] = df2023['steals'] / df2023['games_played']
df2023['bpg'] = df2023['blocks'] / df2023['games_played']
df2023['topg'] = df2023['turnovers'] / df2023['games_played']
df2023['3pg'] = df2023['made_three_point_field_goals'] / df2023['games_played']

print("\n2023 (within top 120)")
print(f'FG%: elite: {df2023["FGperc"].sort_values(ascending=False).head(120).quantile(0.75).round(decimals=3)}, \
      avg: {df2023["FGperc"].sort_values(ascending=False).head(120).quantile(0.5).round(decimals=3)}, poor: {df2023["FGperc"].sort_values(ascending=False).head(120).quantile(0.25).round(decimals=3)}')
print(f'FT%: elite: {df2023["FTperc"].sort_values(ascending=False).head(120).quantile(0.75).round(decimals=3)}, \
      avg: {df2023["FTperc"].sort_values(ascending=False).head(120).quantile(0.5).round(decimals=3)}, poor: {df2023["FTperc"].sort_values(ascending=False).head(120).quantile(0.25).round(decimals=3)}')
print(f'3PTM: elite: {df2023["3pg"].sort_values(ascending=False).head(120).quantile(0.75).round(decimals=3)}, \
      avg: {df2023["3pg"].sort_values(ascending=False).head(120).quantile(0.5).round(decimals=3)}, poor: {df2023["3pg"].sort_values(ascending=False).head(120).quantile(0.25).round(decimals=3)}')
print(f'Points: elite: {df2023["ppg"].sort_values(ascending=False).head(120).quantile(0.75).round(decimals=3)}, \
      avg: {df2023["ppg"].sort_values(ascending=False).head(120).quantile(0.5).round(decimals=3)}, poor: {df2023["ppg"].sort_values(ascending=False).head(120).quantile(0.25).round(decimals=3)}')
print(f'Assists: elite: {df2023["apg"].sort_values(ascending=False).head(120).quantile(0.75).round(decimals=3)}, \
      avg: {df2023["apg"].sort_values(ascending=False).head(120).quantile(0.5).round(decimals=3)}, poor: {df2023["apg"].sort_values(ascending=False).head(120).quantile(0.25).round(decimals=3)}')
print(f'Rebounds: elite: {df2023["rpg"].sort_values(ascending=False).head(120).quantile(0.75).round(decimals=3)}, \
      avg: {df2023["rpg"].sort_values(ascending=False).head(120).quantile(0.5).round(decimals=3)}, poor: {df2023["rpg"].sort_values(ascending=False).head(120).quantile(0.25).round(decimals=3)}')
print(f'Steals: elite: {df2023["spg"].sort_values(ascending=False).head(120).quantile(0.75).round(decimals=3)}, \
      avg: {df2023["spg"].sort_values(ascending=False).head(120).quantile(0.5).round(decimals=3)}, poor: {df2023["spg"].sort_values(ascending=False).head(120).quantile(0.25).round(decimals=3)}')
print(f'Blocks: elite: {df2023["bpg"].sort_values(ascending=False).head(120).quantile(0.75).round(decimals=3)}, \
      avg: {df2023["bpg"].sort_values(ascending=False).head(120).quantile(0.5).round(decimals=3)}, poor: {df2023["bpg"].sort_values(ascending=False).head(120).quantile(0.25).round(decimals=3)}')
print(f'Turnovers: elite: {df2023["topg"].quantile(0.25).round(decimals=3)}, \
      avg: {df2023["topg"].quantile(0.5).round(decimals=3)}, poor: {df2023["topg"].quantile(0.75).round(decimals=3)}')


2023 (within top 120)
FG%: elite: 0.638,       avg: 0.576, poor: 0.542
FT%: elite: 1.0,       avg: 0.894, poor: 0.867
3PTM: elite: 2.647,       avg: 2.109, poor: 1.91
Points: elite: 23.219,       avg: 19.475, poor: 15.352
Assists: elite: 6.102,       avg: 4.823, poor: 3.732
Rebounds: elite: 8.455,       avg: 6.426, poor: 5.462
Steals: elite: 1.318,       avg: 1.089, poor: 0.974
Blocks: elite: 1.033,       avg: 0.783, poor: 0.633
Turnovers: elite: 0.519,       avg: 0.88, poor: 1.5


In [11]:
df2024 = df_pst2024
df2024 = df2024.drop(columns=['positions', 'team'])
df2024 = df2024.groupby(['slug', 'name', 'age'], as_index=False)[['made_field_goals', 'attempted_field_goals', 'made_free_throws', 'attempted_free_throws', \
                         'offensive_rebounds', 'defensive_rebounds', 'games_played', 'points', 'assists', 'steals', \
                            'blocks', 'turnovers', 'made_three_point_field_goals']].sum()
df2024['FGperc'] = df2024['made_field_goals'] / df2024['attempted_field_goals']
df2024['FTperc'] = df2024['made_free_throws'] / df2024['attempted_free_throws']
df2024['rebs'] = df2024['offensive_rebounds'] + df2024['defensive_rebounds']
df2024['ppg'] = df2024['points'] / df2024['games_played']
df2024['rpg'] = df2024['rebs'] / df2024['games_played']
df2024['apg'] = df2024['assists'] / df2024['games_played']
df2024['spg'] = df2024['steals'] / df2024['games_played']
df2024['bpg'] = df2024['blocks'] / df2024['games_played']
df2024['topg'] = df2024['turnovers'] / df2024['games_played']
df2024['3pg'] = df2024['made_three_point_field_goals'] / df2024['games_played']

print("\n2024 (within top 120)")
print(f'FG%: elite: {df2024["FGperc"].sort_values(ascending=False).head(120).quantile(0.75).round(decimals=3)}, \
      avg: {df2024["FGperc"].sort_values(ascending=False).head(120).quantile(0.5).round(decimals=3)}, poor: {df2024["FGperc"].sort_values(ascending=False).head(120).quantile(0.25).round(decimals=3)}')
print(f'FT%: elite: {df2024["FTperc"].sort_values(ascending=False).head(120).quantile(0.75).round(decimals=3)}, \
      avg: {df2024["FTperc"].sort_values(ascending=False).head(120).quantile(0.5).round(decimals=3)}, poor: {df2024["FTperc"].sort_values(ascending=False).head(120).quantile(0.25).round(decimals=3)}')
print(f'3PTM: elite: {df2024["3pg"].sort_values(ascending=False).head(120).quantile(0.75).round(decimals=3)}, \
      avg: {df2024["3pg"].sort_values(ascending=False).head(120).quantile(0.5).round(decimals=3)}, poor: {df2024["3pg"].sort_values(ascending=False).head(120).quantile(0.25).round(decimals=3)}')
print(f'Points: elite: {df2024["ppg"].sort_values(ascending=False).head(120).quantile(0.75).round(decimals=3)}, \
      avg: {df2024["ppg"].sort_values(ascending=False).head(120).quantile(0.5).round(decimals=3)}, poor: {df2024["ppg"].sort_values(ascending=False).head(120).quantile(0.25).round(decimals=3)}')
print(f'Assists: elite: {df2024["apg"].sort_values(ascending=False).head(120).quantile(0.75).round(decimals=3)}, \
      avg: {df2024["apg"].sort_values(ascending=False).head(120).quantile(0.5).round(decimals=3)}, poor: {df2024["apg"].sort_values(ascending=False).head(120).quantile(0.25).round(decimals=3)}')
print(f'Rebounds: elite: {df2024["rpg"].sort_values(ascending=False).head(120).quantile(0.75).round(decimals=3)}, \
      avg: {df2024["rpg"].sort_values(ascending=False).head(120).quantile(0.5).round(decimals=3)}, poor: {df2024["rpg"].sort_values(ascending=False).head(120).quantile(0.25).round(decimals=3)}')
print(f'Steals: elite: {df2024["spg"].sort_values(ascending=False).head(120).quantile(0.75).round(decimals=3)}, \
      avg: {df2024["spg"].sort_values(ascending=False).head(120).quantile(0.5).round(decimals=3)}, poor: {df2024["spg"].sort_values(ascending=False).head(120).quantile(0.25).round(decimals=3)}')
print(f'Blocks: elite: {df2024["bpg"].sort_values(ascending=False).head(120).quantile(0.75).round(decimals=3)}, \
      avg: {df2024["bpg"].sort_values(ascending=False).head(120).quantile(0.5).round(decimals=3)}, poor: {df2024["bpg"].sort_values(ascending=False).head(120).quantile(0.25).round(decimals=3)}')
print(f'Turnovers: elite: {df2024["topg"].quantile(0.25).round(decimals=3)}, \
      avg: {df2024["topg"].quantile(0.5).round(decimals=3)}, poor: {df2024["topg"].quantile(0.75).round(decimals=3)}')


2024 (within top 120)
FG%: elite: 0.649,       avg: 0.576, poor: 0.535
FT%: elite: 1.0,       avg: 0.891, poor: 0.864
3PTM: elite: 2.729,       avg: 2.177, poor: 1.888
Points: elite: 22.903,       avg: 18.546, poor: 14.871
Assists: elite: 5.705,       avg: 4.628, poor: 3.629
Rebounds: elite: 8.321,       avg: 6.321, poor: 5.384
Steals: elite: 1.296,       avg: 1.102, poor: 1.0
Blocks: elite: 1.107,       avg: 0.795, poor: 0.673
Turnovers: elite: 0.4,       avg: 0.767, poor: 1.368


Now we got a brief understanding of what the data looks like, let's start some analytics.  

# 2. Create a Team Roster Auto-Builder

We divide the process it into 2 steps:  
1. Weekly Player Performance Prediction: Average.  
2. Dynamic Programming for team roster selection.  

# 2-1. Weekly Player Performance Prediction: Average

We will first use the average values to determine the "daily" average.  
Then we will muliply by the total games played that week for the player to get the weekly projection (average).  

Get Average 9-cat stat for each player for recent seasons.  
We want the data from years 2021~2024.  
If the player's a rookie, use only the current season's data (year=2024).  

Lets first ask: which records are in df2022 and not in df2021?

In [12]:
df2022[~df2022['name'].isin(df2021['name'])].sort_values(by='points', ascending=False).head(20)


slug              name  age  made_field_goals  \
558  wagnefr01      Franz Wagner   20               456   
205  greenja05       Jalen Green   19               406   
31   barnesc01    Scottie Barnes   20               459   
123  cunnica01   Cade Cunningham   20               429   
383  mobleev01       Evan Mobley   20               422   
381  mitchda01   Davion Mitchell   23               352   
292  joneshe01     Herbert Jones   23               274   
150  duartch01      Chris Duarte   24               268   
259  hylanbo01      Bones Hyland   21               230   
496  sengual01    Alperen Şengün   19               249   
142  dosunay01       Ayo Dosunmu   22               278   
189  giddejo01       Josh Giddey   19               281   
536  thompkl01     Klay Thompson   31               246   
317  kuminjo01  Jonathan Kuminga   19               236   
307  kispeco01     Corey Kispert   22               234   
346   manntr01          Tre Mann   20               231   
102  chrisjo01  Josh Christopher   20               224   
533  thomaca02        Cam Thomas   20               218   
521  suggsja01       Jalen Suggs   20               198   
589  willizi02   Ziaire Williams   20               191   

     attempted_field_goals  made_free_throws  attempted_free_throws  \
558                    975               189                    219   
205                    954               188                    236   
31                     932               158                    215   
123                   1031               142                    168   
383                    830               167                    252   
381                    843                56                     85   
292                    576               137                    163   
150                    621                90                    112   
259                    571               107                    125   
496                    525               165                    232   
142                    535                53                     78   
189                    670                56                     79   
536                    573                46                     51   
317                    460               130                    190   
307                    514                54                     62   
346                    588                65                     82   
102                    500                83                    113   
533                    503                87                    105   
521                    548               126                    163   
589                    424                43                     55   

     offensive_rebounds  defensive_rebounds  games_played  ...    FGperc  \
558                  84                 272            79  ...  0.467692   
205                  33                 193            67  ...  0.425577   
31                  195                 362            74  ...  0.492489   
123                  58                 296            64  ...  0.416101   
383                 143                 427            69  ...  0.508434   
381                  33                 133            75  ...  0.417556   
292                  98                 198            78  ...  0.475694   
150                  41                 185            55  ...  0.431562   
259                  14                 174            69  ...  0.402802   
496                 138                 255            72  ...  0.474286   
142                  27                 187            77  ...  0.519626   
189                  98                 323            54  ...  0.419403   
536                  15                 110            32  ...  0.429319   
317                  54                 180            70  ...  0.513043   
307                  53                 156            77  ...  0.455253   
346                  25                 146            60  ...

155 players in 2022 didn't appear in 2021: wagnefr01 Franz Wagner, thompkl01 Klay Thompson, ...  
They happen to be rookies in 2022 or players that were injured in the previous year (and now recovered).  

In [13]:
df2021[~df2021['name'].isin(df2022['name'])].sort_values(by='points', ascending=False).head(20)

slug                     name  age  made_field_goals  \
527  willizi01          Zion Williamson   20               634   
292  leonaka01            Kawhi Leonard   29               465   
358  murraja01             Jamal Murray   23               378   
456  simmobe01              Ben Simmons   24               325   
508   walljo01                John Wall   30               293   
371   nunnke01            Kendrick Nunn   25               319   
23   bacondw01             Dwayne Bacon   25               292   
532  wisemja01            James Wiseman   19               191   
446  saricda01              Dario Šarić   26               153   
470  sumneed01            Edmond Sumner   25               148   
473  teaguje01              Jeff Teague   32               123   
422  redicjj01              J.J. Redick   36               106   
34   baynear01              Aron Baynes   34               134   
93   cartemi01  Michael Carter-Williams   29               105   
419  randlch01           Chasson Randle   27                92   
170  gasolma01               Marc Gasol   36                88   
449  scottmi01               Mike Scott   32                76   
393  pattepa01        Patrick Patterson   31                72   
376  okafoja01            Jahlil Okafor   25                63   
249  jamesju01             Justin James   24                52   

     attempted_field_goals  made_free_throws  attempted_free_throws  \
527                   1037               369                    529   
292                    908               261                    295   
358                    792               133                    153   
456                    583               176                    287   
508                    726               158                    211   
371                    658                56                     60   
23                     726               145                    176   
532                    368                54                     86   
446                    342                84                     99   
470                    282                68                     83   
473                    284                89                    105   
422                    267                49                     52   
34                     304                29                     41   
93                     270                49                     80   
419                    237                38                     48   
170                    194                36                     50   
449                    211                 8                     12   
393                    165                13                     17   
376                    102                17                     24   
249                    111                21                     36   

     offensive_rebounds  defensive_rebounds  games_played  ...    FGperc  \
527                 167                 274            61  ...  0.611379   
292                  55                 282            52  ...  0.512115   
358                  37                 157            48  ...  0.477273   
456                  93                 324            58  ...  0.557461   
508                  17                 112            40  ...  0.403581   
371                  21                 158            56  ...  0.484802   
23                   31                 193            72  ...  0.402204   
532                  55                 171            39  ...  0.519022   
446                  46                 144            50  ...  0.447368   
470                  18                  76            53  ...  0.524823   
473                  15                  73            55  ...  0.433099   
422                   4                  60            44  ...  0.397004   
34                   87                 186            53  ...  0.440789   
93                   39                 100            31  ...  0.3

90 players in 2021 didn't appear in 2022: willizi01	Zion Williamson, leonaka01	Kawhi Leonard, ...  
If we want the data from year 2021~2024, find the Set Union of those 4 years's player list where we:  
    concat -> groupby -> average (weighted).   

We want a weighted average of year 2021~2024 since year 2024 should weigh more than older years like 2021.  

In [14]:
# Weighted Sum of average stats: (We can add a 'weights' attribute to each years df. )
    # (0.45*2023stat + 0.35*2022stat + 0.2*2021stat) / (0.45*2023gamesplayed + 0.35*2022gamesplayed + 0.2*2021gamesplayed)
# 'slug', 'name', (9-cat weighted sum)... 

# set weights for each year
wts = {"2024": 0.70,
       "2023": 0.15, 
       "2022": 0.10, 
       "2021": 0.05}

df2024['wts'] = wts['2024']
df2023['wts'] = wts['2023']
df2022['wts'] = wts['2022']
df2021['wts'] = wts['2021']

def weighted_average(group, values, games_played, weight):
    tmp=[]
    gp = group[games_played]
    wt = group[weight]
    for value in values:
        val = group[value]   
        if value in ['FGperc', 'FTperc']:    
            tmp.append((val * wt).sum())
        else:
            tmp.append((val * wt).sum() / (gp*wt).sum())
    return pd.Series(tmp, index=values)

values=['FGperc', 'FTperc', 'made_three_point_field_goals', 'points', 'assists', 'rebs', 'steals', 'blocks', 'turnovers']

# Although outer join is a little faster, it will miss the rows with the "same exact values". Concatenation wouldn't miss.
df_pred_avg_21_24 = pd.concat([df2021, df2022, df2023, df2024]).fillna(0)[['slug', 'name', *values, 'games_played', 'wts']] \
    .groupby(['slug', 'name'], as_index=False) \
    .apply(weighted_average, values, 'games_played', 'wts')


Let's have a look at the results of famous players.  

In [15]:
df_pred_avg_21_24[df_pred_avg_21_24['name'].isin(['Klay Thompson', 'Stephen Curry', 'Kevin Durant', 'LeBron James', 'Nikola Jokić', \
                            'Giannis Antetokounmpo', 'Tyrese Haliburton', 'De\'Aaron Fox', 'Victor Wembanyama'])]

slug                   name    FGperc    FTperc  \
15   antetgi01  Giannis Antetokounmpo  0.592618  0.677348   
175  curryst01          Stephen Curry  0.453927  0.922213   
215  duranke01           Kevin Durant  0.529262  0.890425   
247    foxde01           De'Aaron Fox  0.477009  0.734482   
298  halibty01      Tyrese Haliburton  0.492497  0.865417   
383  jamesle01           LeBron James  0.519557  0.735344   
402  jokicni01           Nikola Jokić  0.588234  0.820364   
762  thompkl01          Klay Thompson  0.405279  0.853722   
818  wembavi01      Victor Wembanyama  0.321375  0.563636   

     made_three_point_field_goals     points    assists       rebs    steals  \
15                       0.617204  30.776344   5.868817  11.487097  1.167742   
175                      4.633979  27.424173   5.428734   4.814139  0.970353   
215                      2.141711  29.006684   5.716578   6.548128  0.822193   
247                      2.482204  26.407577   5.926521   4.067738  1.425947   
298                      2.961995  20.801663  10.776722   3.985748  1.349169   
383                      2.239286  26.288095   7.236905   7.553571  1.276190   
402                      1.029970  25.594406   9.027972  12.102897  1.244755   
762                      3.611995  18.686659   2.390453   3.703794  0.544676   
818                      1.441176  19.617647   2.970588  10.294118  1.147059   

       blocks  turnovers  
15   1.086022   3.624731  
175  0.369441   3.128848  
215  1.160428   3.355615  
247  0.376579   2.639495  
298  0.604513   2.497625  
383  0.667857   3.369048  
402  0.846154   3.201798  
762  0.490820   1.670747  
818  3.176471   3.117647

However, these are "daily" stats rather than "weekly".  

**Weekly Projection for All Players:**  
Because we compete against a new opponent every week, we need a player's weekly performance prediction.  
Therefore, we need to take game schedule in to consideration.  
- Weekly(Monday~Sunday) Projection = Daily Prediction(use 'Average' for now) * games played   

We will create a function getWeekProjection_2324() at the end of this subsection.  

First get the 2023-2024 season schedule from bball ref.   

In [16]:
from basketball_reference_web_scraper import client
df_schedule2024 = pd.DataFrame.from_dict(client.season_schedule(season_end_year=2024))
df_schedule2024.head()

start_time                away_team  \
0 2023-10-24 23:30:00+00:00  Team.LOS_ANGELES_LAKERS   
1 2023-10-25 02:00:00+00:00        Team.PHOENIX_SUNS   
2 2023-10-25 23:00:00+00:00     Team.HOUSTON_ROCKETS   
3 2023-10-25 23:00:00+00:00      Team.BOSTON_CELTICS   
4 2023-10-25 23:00:00+00:00  Team.WASHINGTON_WIZARDS   

                    home_team  away_team_score  home_team_score  
0         Team.DENVER_NUGGETS            107.0            119.0  
1  Team.GOLDEN_STATE_WARRIORS            108.0            104.0  
2          Team.ORLANDO_MAGIC             86.0            116.0  
3        Team.NEW_YORK_KNICKS            108.0            104.0  
4         Team.INDIANA_PACERS            120.0            143.0

Get Player List of 2024 and join with the current team name.  
Traded Players have multiple rows in df_pst2024.  
For example, James Harden was traded at the beginning of the season so he has two records: one with the 76ers, and one with the Clippers.  

We only need to know the newest team of that player.  

In [17]:
# df_pst2024 has 489, but only 483 unique names
print(df_pst2024.shape)
print(len(df_pst2024['name'].unique()))

# find the duplicate names ==> aka players that've been traded this season
df_tmp = df_pst2024[['name']].groupby(['name'], as_index=False).size()
traded_playerlist = list(df_tmp[df_tmp['size']>1]['name'])

df_pst2024[df_pst2024['name'].isin(traded_playerlist)]


(537, 22)
523


slug                name                  positions  age  \
0    achiupr01    Precious Achiuwa          [Position.CENTER]   24   
1    achiupr01    Precious Achiuwa   [Position.POWER_FORWARD]   24   
13   anunoog01          OG Anunoby   [Position.SMALL_FORWARD]   26   
14   anunoog01          OG Anunoby   [Position.SMALL_FORWARD]   26   
31   barrerj01          RJ Barrett  [Position.SHOOTING_GUARD]   23   
32   barrerj01          RJ Barrett  [Position.SHOOTING_GUARD]   23   
36   batumni01       Nicolas Batum   [Position.POWER_FORWARD]   35   
37   batumni01       Nicolas Batum   [Position.POWER_FORWARD]   35   
104  covinro01    Robert Covington   [Position.POWER_FORWARD]   33   
105  covinro01    Robert Covington   [Position.SMALL_FORWARD]   33   
141  flynnma01       Malachi Flynn     [Position.POINT_GUARD]   25   
142  flynnma01       Malachi Flynn     [Position.POINT_GUARD]   25   
280  loftoke01  Kenneth Lofton Jr.   [Position.POWER_FORWARD]   21   
281  loftoke01  Kenneth Lofton Jr.   [Position.POWER_FORWARD]   21   
298  martike04           KJ Martin   [Position.SMALL_FORWARD]   23   
299  martike04           KJ Martin   [Position.SMALL_FORWARD]   23   
304   mayssk01         Skylar Mays     [Position.POINT_GUARD]   26   
305   mayssk01         Skylar Mays  [Position.SHOOTING_GUARD]   26   
370  petrufi01      Filip Petrušev          [Position.CENTER]   23   
371  petrufi01      Filip Petrušev          [Position.CENTER]   23   
394  quickim01   Immanuel Quickley  [Position.SHOOTING_GUARD]   24   
395  quickim01   Immanuel Quickley     [Position.POINT_GUARD]   24   
463  theisda01        Daniel Theis          [Position.CENTER]   31   
464  theisda01        Daniel Theis          [Position.CENTER]   31   
479  tuckepj01         P.J. Tucker   [Position.POWER_FORWARD]   38   
480  tuckepj01         P.J. Tucker   [Position.POWER_FORWARD]   38   
527  windldy01       Dylan Windler   [Position.SMALL_FORWARD]   27   
528  windldy01       Dylan Windler   [Position.SMALL_FORWARD]   27   

                            team  games_played  games_started  minutes_played  \
0           Team.TORONTO_RAPTORS            25              0             437   
1           Team.NEW_YORK_KNICKS             8              0             102   
13          Team.TORONTO_RAPTORS            27             27             900   
14          Team.NEW_YORK_KNICKS             8              8             282   
31          Team.NEW_YORK_KNICKS            26             26             766   
32          Team.TORONTO_RAPTORS             8              8             264   
36     Team.LOS_ANGELES_CLIPPERS             3              0              54   
37       Team.PHILADELPHIA_76ERS            25             21             638   
104    Team.LOS_ANGELES_CLIPPERS             3              3              69   
105      Team.PHILADELPHIA_76ERS            26              3             419   
141         Team.TORONTO_RAPTORS            31              0             474   
142         Team.NEW_YORK_KNICKS             5              0              28   
280       Team.MEMPHIS_GRIZZLIES            15              0              99   
281      Team.PHILADELPHIA_76ERS             1              0               4   
298    Team.LOS_ANGELES_CLIPPERS             2              0              31   
299      Team.PHILADELPHIA_76ERS            24              0             145   
304  Team.PORTLAND_TRAIL_BLAZERS            21              5             356   
305      Team.LOS_ANGELES_LAKERS             2              0              22   
370      Team.PHILADELPHIA_76ERS             1              0               3   
371        Team.SACRAMENTO_KINGS             2              0               7   
394         Team.NEW_YORK_KNICKS            30              0             720   
395         Team.TORONTO_RAPTORS             8              8             248   
463          Team.INDIANA_PACERS             1              0               8   
464    Team.LOS_ANGELES_CLIPPERS   

Conclusion from above cell: rows that appear later (larger index) are the newer teams of that traded player.  

In [18]:

playerlist2024 = df_pst2024[['slug', 'name', 'team']]

playerlist2024['idx'] = range(0,len(playerlist2024))

# only keep the row with the larger index if rows having same names and slugs
playerlist2024 = playerlist2024.loc[playerlist2024.groupby(['slug', 'name'])['idx'].idxmax()].drop(columns=['idx'])
playerlist2024


C:\Users\joe_c\AppData\Local\Temp\ipykernel_24568\446719571.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  playerlist2024['idx'] = range(0,len(playerlist2024))


slug                      name                         team
1    achiupr01          Precious Achiuwa         Team.NEW_YORK_KNICKS
2    adebaba01               Bam Adebayo              Team.MIAMI_HEAT
3    agbajoc01              Ochai Agbaji               Team.UTAH_JAZZ
4    aldamsa01              Santi Aldama       Team.MEMPHIS_GRIZZLIES
5    alexani01  Nickeil Alexander-Walker  Team.MINNESOTA_TIMBERWOLVES
..         ...                       ...                          ...
532  youngth01            Thaddeus Young         Team.TORONTO_RAPTORS
533  youngtr01                Trae Young           Team.ATLANTA_HAWKS
534  yurtsom01            Omer Yurtseven               Team.UTAH_JAZZ
535  zelleco01               Cody Zeller    Team.NEW_ORLEANS_PELICANS
536  zubaciv01               Ivica Zubac    Team.LOS_ANGELES_CLIPPERS

[523 rows x 3 columns]

Now let's add the team information to the player projection dataframe by inner-joining these two dataframes:
1. playerlist2024 (team information)
2. df_pred_avg_21_24 (player projection)

In [19]:
df_weeklyproj_avg_24 = playerlist2024.merge(df_pred_avg_21_24, on=['slug', 'name'], how='inner')
df_weeklyproj_avg_24


slug                      name                         team  \
0    achiupr01          Precious Achiuwa         Team.NEW_YORK_KNICKS   
1    adebaba01               Bam Adebayo              Team.MIAMI_HEAT   
2    agbajoc01              Ochai Agbaji               Team.UTAH_JAZZ   
3    aldamsa01              Santi Aldama       Team.MEMPHIS_GRIZZLIES   
4    alexani01  Nickeil Alexander-Walker  Team.MINNESOTA_TIMBERWOLVES   
..         ...                       ...                          ...   
518  youngth01            Thaddeus Young         Team.TORONTO_RAPTORS   
519  youngtr01                Trae Young           Team.ATLANTA_HAWKS   
520  yurtsom01            Omer Yurtseven               Team.UTAH_JAZZ   
521  zelleco01               Cody Zeller    Team.NEW_ORLEANS_PELICANS   
522  zubaciv01               Ivica Zubac    Team.LOS_ANGELES_CLIPPERS   

       FGperc    FTperc  made_three_point_field_goals     points    assists  \
0    0.465046  0.601108                      0.473621   7.460432   1.200240   
1    0.518206  0.781657                      0.023143  20.764921   3.892814   
2    0.367244  0.657033                      1.104575   6.406536   0.994771   
3    0.418618  0.614216                      1.311888   9.398601   1.464336   
4    0.414322  0.697632                      1.347052   7.224694   2.269188   
..        ...       ...                           ...        ...        ...   
518  0.522892  0.415534                      0.155378   5.304781   2.083665   
519  0.428053  0.866316                      2.941176  27.174370  10.448529   
520  0.491520  0.607295                      0.111579   4.016842   0.602105   
521  0.450095  0.640365                      0.033395   3.434137   1.102041   
522  0.648683  0.670116                      0.001016  11.436992   1.253049   

          rebs    steals    blocks  turnovers  
0     5.350120  0.526379  0.525180   1.052758  
1    10.059683  1.136419  0.904994   2.545676  
2     2.505882  0.501961  0.479739   0.691503  
3     5.050350  0.577622  0.660140   0.923077  
4     1.972191  0.806452  0.519466   1.043382  
..         ...       ...       ...        ...  
518   3.276892  0.936255  0.209163   0.926295  
519   3.199580  1.238445  0.188025   4.150210  
520   4.418947  0.202105  0.391579   0.846316  
521   3.480519  0.200371  0.153989   0.528757  
522   9.405488  0.299797  1.258130   1.282520  

[523 rows x 12 columns]

Now we want to know how many games does each team play this week?  

Let's get the 2023-2024 NBA game schedule.  
We also want to add a 'week' column for later grouping.  

In [20]:
import datetime

# first assume every player on the team plays. Just get the weekly schedule of the player's team. 
# view schema: team, week (starting from 2023-10-24), gamesplayed (during that week) 
# can also add columns: homegames, awaygames

df_schedule2024.start_time = df_schedule2024['start_time'].dt.tz_convert('America/Los_Angeles')
df_schedule2024.start_time = df_schedule2024.start_time.dt.date

# Create 'week' column: 
# week1: 2023-10-24 ~ 10-29, week2: 10-30 ~ 11-5, week3: 11-6 ~ 11-12, ...
# R dplyr::mutate, equivalent to pandas conditionally with list comprehension, or apply lambda

seasonStartDate = df_schedule2024.start_time[0]
seasonStartDate_weekday = datetime.timedelta(seasonStartDate.weekday())
df_schedule2024['week'] = df_schedule2024['start_time'].apply(lambda x: (x - seasonStartDate + seasonStartDate_weekday).days//7 + 1)
df_schedule2024


start_time                    away_team                    home_team  \
0     2023-10-24      Team.LOS_ANGELES_LAKERS          Team.DENVER_NUGGETS   
1     2023-10-24            Team.PHOENIX_SUNS   Team.GOLDEN_STATE_WARRIORS   
2     2023-10-25         Team.HOUSTON_ROCKETS           Team.ORLANDO_MAGIC   
3     2023-10-25          Team.BOSTON_CELTICS         Team.NEW_YORK_KNICKS   
4     2023-10-25      Team.WASHINGTON_WIZARDS          Team.INDIANA_PACERS   
...          ...                          ...                          ...   
1226  2024-04-14            Team.PHOENIX_SUNS  Team.MINNESOTA_TIMBERWOLVES   
1227  2024-04-14      Team.LOS_ANGELES_LAKERS    Team.NEW_ORLEANS_PELICANS   
1228  2024-04-14        Team.DALLAS_MAVERICKS   Team.OKLAHOMA_CITY_THUNDER   
1229  2024-04-14  Team.PORTLAND_TRAIL_BLAZERS        Team.SACRAMENTO_KINGS   
1230  2024-04-14         Team.DETROIT_PISTONS       Team.SAN_ANTONIO_SPURS   

      away_team_score  home_team_score  week  
0               107.0            119.0     1  
1               108.0            104.0     1  
2                86.0            116.0     1  
3               108.0            104.0     1  
4               120.0            143.0     1  
...               ...              ...   ...  
1226              NaN              NaN    25  
1227              NaN              NaN    25  
1228              NaN              NaN    25  
1229              NaN              NaN    25  
1230              NaN              NaN    25  

[1231 rows x 6 columns]

Group by week and team to know how many games does each team play this week?   

In [21]:
# groupby week and team
df_gamesplayed_weekly = df_schedule2024.copy()
df_gamesplayed_weekly['away_team'] = df_gamesplayed_weekly['away_team'].apply(lambda x: x.value)
df_gamesplayed_weekly['home_team'] = df_gamesplayed_weekly['home_team'].apply(lambda x: x.value)
df_gamesplayed_weekly

df_tmp1 = df_gamesplayed_weekly[['week', 'away_team']].groupby(['week', 'away_team'], as_index=False).size()
df_tmp2 = df_gamesplayed_weekly[['week', 'home_team']].groupby(['week', 'home_team'], as_index=False).size()
df_tmp1.rename(columns={'away_team':"team"}, inplace=True)
df_tmp2.rename(columns={'home_team':"team"}, inplace=True)
df_gamesplayed_weekly = pd.concat([df_tmp1, df_tmp2]).groupby(['week', 'team'], as_index=False).sum()
# df_gamesplayed_weekly.sort_values(by=['week', 'size'], ascending = [True, False])
df_gamesplayed_weekly


week                team  size
0       1       ATLANTA HAWKS     3
1       1      BOSTON CELTICS     2
2       1       BROOKLYN NETS     2
3       1   CHARLOTTE HORNETS     2
4       1       CHICAGO BULLS     3
..    ...                 ...   ...
745    25    SACRAMENTO KINGS     4
746    25   SAN ANTONIO SPURS     4
747    25     TORONTO RAPTORS     4
748    25           UTAH JAZZ     4
749    25  WASHINGTON WIZARDS     3

[750 rows x 3 columns]

Example: De'Aaron Fox's week6 projection  

In [22]:
# De'Aaron Fox week6 projection
week = 13
print(type(playerlist2024[playerlist2024.name=='De\'Aaron Fox'].team.values[0]))
team = playerlist2024[playerlist2024.name=='De\'Aaron Fox'].team.values[0].value
gamesPlayed_week = df_gamesplayed_weekly[(df_gamesplayed_weekly['week']==week) & (df_gamesplayed_weekly['team']==team)]['size'].iloc[0]
print(df_weeklyproj_avg_24[df_weeklyproj_avg_24['name']=='De\'Aaron Fox'].iloc[:,3:5]) # Percentage doesn't need to multiply by games played. 
df_weeklyproj_avg_24[df_weeklyproj_avg_24['name']=='De\'Aaron Fox'].iloc[:,5:] * gamesPlayed_week


<enum 'Team'>
       FGperc    FTperc
141  0.477009  0.734482


made_three_point_field_goals     points    assists      rebs    steals  \
141                      4.964409  52.815155  11.853042  8.135476  2.851894   

       blocks  turnovers  
141  0.753157    5.27899

Create a function getWeekProjection_2324()  
The input is an nba week number, and outputs are the weekly projection for all the players in that specified week.  
We have a look at the outputs sorted by made_three_point_field_goals.   

In [23]:
# func(week) ==> weekly projection of every player
def current_week():
    seasonStartDate = df_schedule2024.start_time[0]
    seasonStartDate_weekday = datetime.timedelta(seasonStartDate.weekday())
    week = (datetime.date.today() - seasonStartDate + seasonStartDate_weekday).days//7 + 1
    return week

def getWeekProjection_2324(week=current_week()):
    df_weeklyproj_avg_24 = playerlist2024.merge(df_pred_avg_21_24, on=['slug', 'name'], how='inner')
    gamesPlayed_week = df_gamesplayed_weekly[(df_gamesplayed_weekly['week']==week) & (df_gamesplayed_weekly['team']==team)]['size'].iloc[0]
    df_weeklyproj_avg_24.iloc[:,5:] = df_weeklyproj_avg_24.iloc[:,5:] * gamesPlayed_week
    return df_weeklyproj_avg_24

df_tmp = getWeekProjection_2324(current_week())
df_tmp.sort_values(by='made_three_point_field_goals', ascending=False).head(20)




slug               name                         team    FGperc  \
105  curryst01      Stephen Curry   Team.GOLDEN_STATE_WARRIORS  0.453927   
455  thompkl01      Klay Thompson   Team.GOLDEN_STATE_WARRIORS  0.405279   
119  doncilu01        Luka Dončić        Team.DALLAS_MAVERICKS  0.483564   
151  georgpa01        Paul George    Team.LOS_ANGELES_CLIPPERS  0.461962   
269  lillada01     Damian Lillard         Team.MILWAUKEE_BUCKS  0.432792   
320  mitchdo01   Donovan Mitchell     Team.CLEVELAND_CAVALIERS  0.460798   
21    ballla01        LaMelo Ball       Team.CHARLOTTE_HORNETS  0.430451   
298  mccolcj01        CJ McCollum    Team.NEW_ORLEANS_PELICANS  0.461135   
426  simonan01    Anfernee Simons  Team.PORTLAND_TRAIL_BLAZERS  0.425345   
472  vanvlfr01      Fred VanVleet         Team.HOUSTON_ROCKETS  0.402412   
447  tatumja01       Jayson Tatum          Team.BOSTON_CELTICS  0.471924   
177  hardati02   Tim Hardaway Jr.        Team.DALLAS_MAVERICKS  0.417317   
195  hieldbu01        Buddy Hield          Team.INDIANA_PACERS  0.438300   
24    banede01       Desmond Bane       Team.MEMPHIS_GRIZZLIES  0.473368   
216  irvinky01       Kyrie Irving        Team.DALLAS_MAVERICKS  0.482087   
35   beaslma01      Malik Beasley         Team.MILWAUKEE_BUCKS  0.467099   
46   bogdabo01  Bogdan Bogdanović           Team.ATLANTA_HAWKS  0.429572   
194  herroty01        Tyler Herro              Team.MIAMI_HEAT  0.440483   
175  halibty01  Tyrese Haliburton          Team.INDIANA_PACERS  0.492497   
287  markkla01    Lauri Markkanen               Team.UTAH_JAZZ  0.491280   

       FTperc  made_three_point_field_goals     points    assists       rebs  \
105  0.922213                      9.267959  54.848347  10.857469   9.628278   
455  0.853722                      7.223990  37.373317   4.780906   7.407589   
119  0.766711                      6.986207  64.197701  17.572414  16.744828   
151  0.897515                      6.796569  47.801471   8.504902  11.644608   
269  0.916564                      6.788250  53.767442  14.031824   8.832313   
320  0.872962                      6.687578  55.279799  10.624843   9.731493   
21   0.861924                      6.493601  44.603291  15.462523  12.164534   
298  0.776286                      6.413437  41.144703  10.485788   8.607235   
426  0.897340                      6.389610  39.675325   8.233766   5.821429   
472  0.871057                      6.304017  36.343105  15.611292   7.904452   
447  0.826492                      6.257322  55.470711   8.983264  16.776151   
177  0.844487                      6.253615  33.664071   3.575083   7.176863   
195  0.852965                      6.242014  28.778316   5.320426   7.779284   
24   0.868578                      6.125000  43.399123   8.859649   9.028509   
216  0.896224                      6.080997  52.778816  10.987539   9.928349   
35   0.764551                      6.026804  24.839175   2.927835   7.465979   
46   0.883463                      6.023529  32.155294   5.545882   6.637647   
194  0.859347                      5.967695  41.861968   8.334802  10.784141   
175  0.865417                      5.923990  41.603325  21.553444   7.971496   
287  0.871800                      5.907204  44.737485   3.389499  16.156288   

       steals    blocks  turnovers  
105  1.940707  0.738883   6.257697  
455  1.089351  0.981640   3.341493  
119  2.625287  1.087356   7.928736  
151  3.218137  0.715686   5.095588  
269  1.904529  0.433293   5.493268  
320  3.340025  0.757842   5.573400  
21   3.155393  0.610603   7.166362  
298  2.284238  1.186047   3.726098  
426  1.116883  0.292208   4.133117  
472  2.657980  1.361564   3.711183  
447  2.043933  1.146444   5.531381  
177  1.063404  0.162403   1.753059  
195  1.752178  0.983543   3.142304  
24   2.028509  0.921053   4.484649  
216  2.591900  1.165109   3.844237  
35   1.478351  0.263918   1.579381  
46   2.484706  0.487059   2.200000  
194  1.879589  0.320117   4.848752  
175  2.698337  1.209026   4.9

Notes: In the 2023-2024 season, week7 is the week of In-Season Tournament that won't been scheduled beforehand. 

# 2-2. Dynamic Programming for team roster selection.  

13-roster Team Selection Mechanism under a 12-team Snake Draft.   
Create the 13-roster Team from the Weekly Projection using similar techinques as when solving the 0/1 knapsack problem.  

Find the best 13 players to win a 9-cat game regardless of the position constraint and the snake draft constraint.  

1. Brute Force: 13 players out of 483 (mylist) or 548 (fantasypros draft list). Using recursive has an O(2^m) time complexity, and C(483, 13) = 1.062392279 E+25 is too large of a number.   

2. Dynamic Programming: 0/1 Knapsack problem  
    Given N **items** where each item has some weight and **profit** associated with it and also given a bag with **capacity** W, [i.e., the bag can hold at most W weight in it]. The goal is to put the items into the bag such that is has the **max sum of profits** possible.  
    
    N: total player list (**items**)  
    W: 13-player roster team that consists of 1PG, 1SG, 1G, 1SF, 1PF, 1F, 2C, 2Util, 3BN (**capacity**)  
    profit: 9-cat stats (**profits**)  
    goal: categorical wins (**max sum of profits**)  




Let's first solve 0/1 knapsack problem.  

Assume that the 13-player team doesn't have position constraints.  
aka. every player's weight is 1, where capacity W=13.  

The difference in our case is that: **"profit" is multidimensional** since they're 9-cat stats.  
==> we can do an "operation overload" for this multidimensional-categorical profit  
    usually: profit1=8, profit2=4, profit3=2;  
        profit1 = 8 > profit2 + profit3 = 6  
    now: profit1 = [0,2,2,2,2,2,2,2,0], profit2 = [100,0,0,0,0,0,0,1,1], profit3 = [0,100,0,0,0,0,1,0,0]   
        profit1 = [0,2,2,2,2,2,2,2,2] > profit2 + profit3 = [100,100,0,0,0,0,1,1,1]   
        profit1 wins the last 7 categories over profit2 + profit3's first 2 categories. 
        list-elementwise greaterthan.   

In [24]:
# define the new greaterthan
def profit_greaterthan(A, B): 
    # A and B are lists. 
    l = []
    for (a, b) in zip(A,B):
        if a > b:
            l.append(1)
        elif a==b:
            l.append(0)
        else:
            l.append(-1)
    if sum(l) > 0:
        return True
    else:
        return False

def profit_max(A_exclude, B_include): 
    # A, B, and idx are lists. 
    if profit_greaterthan(A_exclude, B_include):
        return A_exclude
    else:
        return B_include
    
def profit_addition(A, B):
    # A and B are lists. 
    return [sum(x) for x in zip(A, B)]


# Space optimized Approach (code is inspired by Prosun Kumar Sarkar)
def knapSack(W, wt, val, n): 
    # Making the dp array each with 9 category values initialized as zero.  
    dp = [[0]*9 for i in range(W+1)]

    # Taking first i elements 
    for i in range(1, n+1): 
        # Starting from back, 
        # so that we also have data of 
        # previous computation when taking i-1 items 
        for w in range(W, 0, -1): 
            if wt[i-1] <= w: 
                # Find the max value 
                dp[w] = profit_max(dp[w], profit_addition(dp[w-wt[i-1]], val[i-1])) 
      
    # Returning the maximum value of knapsack 
    return dp[W]
  
  
profit = [[100,0,0,0,0,0,0,1,1], [0,100,0,0,0,0,1,0,0], [0,2,2,2,2,2,2,2,0]]
weight = [1, 1, 1]
W = 2
n = len(profit)
print(knapSack(W, weight, profit, n))


profit = [1, 7, 4]
weight = [3, 8, 2]
W = 6
N = 3





[100, 2, 2, 2, 2, 2, 2, 3, 1]


However, the above doesn't tell us which items were selected.  
We need to output the selected indexes.  

In [25]:
# define the new greaterthan
def profit_greaterthan(A, B): 
    # A and B are lists. 
    l = []
    for (a, b) in zip(A,B):
        if a > b:
            l.append(1)
        elif a==b:
            l.append(0)
        else:
            l.append(-1)
    if sum(l) > 0:
        return True
    else:
        return False

def profit_max(A_exclude, B_include): 
    # A, B, and idx are lists. 
    if profit_greaterthan(A_exclude, B_include):
        return A_exclude, 0
    else:
        return B_include, 1
    
def profit_addition(A, B):
    # A and B are lists. 
    return [sum(x) for x in zip(A, B)]


# Space optimized Approach (code is inspired by Prosun Kumar Sarkar)
def knapSack(W, wt, val, n): 
    # Making the dp array each with 9 category values initialized as zero.  
    dp = dp = [[[0]*9 if j==0 else [] for j in range(2)] for i in range(W+1)]

    # Taking first i elements 
    for i in range(1, n+1): 
        # Starting from back, 
        # so that we also have data of 
        # previous computation when taking i-1 items 
        for w in range(W, 0, -1): 
            if wt[i-1] <= w: 
                # Find the max value 
                dp[w][0], left_or_right = profit_max(dp[w][0], profit_addition(dp[w-wt[i-1]][0], val[i-1])) 
                if left_or_right==1: #include
                    dp[w][1] = dp[w-wt[i-1]][1].copy()
                    dp[w][1].append(i-1) 
                    
    # Returning the maximum value of knapsack 
    return dp[W]
  
  
profit = [[0,2,2,2,2,2,2,2,0], [0,100,0,0,0,0,1,0,0], [100,0,0,0,0,0,0,1,1]]
weight = [1, 1, 1]
W = 2
n = len(profit)
print(knapSack(W, weight, profit, n))



[[100, 2, 2, 2, 2, 2, 2, 3, 1], [0, 2]]


Try on our nba data df_currentweekproj = getWeekProjection_2324(current_week()). 

In [26]:
df_currentweekproj = getWeekProjection_2324(current_week())
df_currentweekproj

slug                      name                         team  \
0    achiupr01          Precious Achiuwa         Team.NEW_YORK_KNICKS   
1    adebaba01               Bam Adebayo              Team.MIAMI_HEAT   
2    agbajoc01              Ochai Agbaji               Team.UTAH_JAZZ   
3    aldamsa01              Santi Aldama       Team.MEMPHIS_GRIZZLIES   
4    alexani01  Nickeil Alexander-Walker  Team.MINNESOTA_TIMBERWOLVES   
..         ...                       ...                          ...   
518  youngth01            Thaddeus Young         Team.TORONTO_RAPTORS   
519  youngtr01                Trae Young           Team.ATLANTA_HAWKS   
520  yurtsom01            Omer Yurtseven               Team.UTAH_JAZZ   
521  zelleco01               Cody Zeller    Team.NEW_ORLEANS_PELICANS   
522  zubaciv01               Ivica Zubac    Team.LOS_ANGELES_CLIPPERS   

       FGperc    FTperc  made_three_point_field_goals     points    assists  \
0    0.465046  0.601108                      0.947242  14.920863   2.400480   
1    0.518206  0.781657                      0.046285  41.529842   7.785627   
2    0.367244  0.657033                      2.209150  12.813072   1.989542   
3    0.418618  0.614216                      2.623776  18.797203   2.928671   
4    0.414322  0.697632                      2.694105  14.449388   4.538376   
..        ...       ...                           ...        ...        ...   
518  0.522892  0.415534                      0.310757  10.609562   4.167331   
519  0.428053  0.866316                      5.882353  54.348739  20.897059   
520  0.491520  0.607295                      0.223158   8.033684   1.204211   
521  0.450095  0.640365                      0.066790   6.868275   2.204082   
522  0.648683  0.670116                      0.002033  22.873984   2.506098   

          rebs    steals    blocks  turnovers  
0    10.700240  1.052758  1.050360   2.105516  
1    20.119367  2.272838  1.809988   5.091352  
2     5.011765  1.003922  0.959477   1.383007  
3    10.100699  1.155245  1.320280   1.846154  
4     3.944383  1.612903  1.038932   2.086763  
..         ...       ...       ...        ...  
518   6.553785  1.872510  0.418327   1.852590  
519   6.399160  2.476891  0.376050   8.300420  
520   8.837895  0.404211  0.783158   1.692632  
521   6.961039  0.400742  0.307978   1.057514  
522  18.810976  0.599593  2.516260   2.565041  

[523 rows x 12 columns]

In [27]:
df_profit = df_currentweekproj.iloc[:,3:]
df_profit.turnovers = df_profit.turnovers*-1
profit = df_profit.values.tolist()
n = len(profit)
weight = [1]*n
W = 13

values, player_idx = knapSack(W, weight, profit, n)
print(values)
print(player_idx)

df_currentweekproj.iloc[player_idx]

[6.417129900317337, 11.139393042598595, 60.50462082438675, 667.4448631979632, 171.58920191222305, 178.87094353861724, 32.63981330038111, 23.01786839189976, -71.74374157884779]
[107, 119, 125, 131, 151, 155, 175, 216, 236, 447, 472, 496, 519]


slug                     name                        team    FGperc  \
107  davisan02            Anthony Davis     Team.LOS_ANGELES_LAKERS  0.547823   
119  doncilu01              Luka Dončić       Team.DALLAS_MAVERICKS  0.483564   
125  duranke01             Kevin Durant           Team.PHOENIX_SUNS  0.529262   
131  embiijo01              Joel Embiid     Team.PHILADELPHIA_76ERS  0.534460   
151  georgpa01              Paul George   Team.LOS_ANGELES_CLIPPERS  0.461962   
155  gilgesh01  Shai Gilgeous-Alexander  Team.OKLAHOMA_CITY_THUNDER  0.533721   
175  halibty01        Tyrese Haliburton         Team.INDIANA_PACERS  0.492497   
216  irvinky01             Kyrie Irving       Team.DALLAS_MAVERICKS  0.482087   
236  jokicni01             Nikola Jokić         Team.DENVER_NUGGETS  0.588234   
447  tatumja01             Jayson Tatum         Team.BOSTON_CELTICS  0.471924   
472  vanvlfr01            Fred VanVleet        Team.HOUSTON_ROCKETS  0.402412   
496  whitede01            Derrick White         Team.BOSTON_CELTICS  0.461132   
519  youngtr01               Trae Young          Team.ATLANTA_HAWKS  0.428053   

       FTperc  made_three_point_field_goals     points    assists       rebs  \
107  0.795329                      0.734940  49.877108   6.612048  23.653012   
119  0.766711                      6.986207  64.197701  17.572414  16.744828   
125  0.890425                      4.283422  58.013369  11.433155  13.096257   
131  0.869830                      2.356511  66.740834  10.111252  22.328698   
151  0.897515                      6.796569  47.801471   8.504902  11.644608   
155  0.881497                      2.347826  60.015608  12.066890  10.740245   
175  0.865417                      5.923990  41.603325  21.553444   7.971496   
216  0.896224                      6.080997  52.778816  10.987539   9.928349   
236  0.820364                      2.059940  51.188811  18.055944  24.205794   
447  0.826492                      6.257322  55.470711   8.983264  16.776151   
472  0.871057                      6.304017  36.343105  15.611292   7.904452   
496  0.892216                      4.490526  29.065263   9.200000   7.477895   
519  0.866316                      5.882353  54.348739  20.897059   6.399160   

       steals    blocks  turnovers  
107  2.356627  4.739759   4.491566  
119  2.625287  1.087356   7.928736  
125  1.644385  2.320856   6.711230  
131  2.149178  3.506953   7.127686  
151  3.218137  0.715686   5.095588  
155  3.872910  1.658863   4.579710  
175  2.698337  1.209026   4.995249  
216  2.591900  1.165109   3.844237  
236  2.489510  1.692308   6.403596  
447  2.043933  1.146444   5.531381  
472  2.657980  1.361564   3.711183  
496  1.814737  2.037895   3.023158  
519  2.476891  0.376050   8.300420

EXPERIMENT: "choosing the top 13 players" vs. "using knapsack algorithm".  

In [28]:
# EXPERIMENT
# try simple sorting. getting the top 13. 

df_profit = df_currentweekproj.iloc[:,3:]
df_profit.turnovers = df_profit.turnovers*-1
profit = df_profit.values.tolist()

def function_top13(P):
    top13 = []
    visited = []
    for i in range(13):
        max_p = [0]*9
        max_idx = -1
        for j, p in enumerate(P):
            if j in visited:
                continue
            if profit_greaterthan(p, max_p):
                max_p = p
                max_idx = j
        visited.append(max_idx)
        top13.append(max_idx)
    return top13

# without using knapsack algo (Top 13).
tmp1 = df_currentweekproj.iloc[function_top13(profit)]
profit_simple = [0]*7
for p in tmp1.iloc[:, 3:].values.tolist():
    profit_simple = profit_addition(profit_simple, p)

# with using knapsack algo.
tmp2 = df_currentweekproj.iloc[player_idx]
profit_knapsack = [0]*7
for p in tmp2.iloc[:, 3:].values.tolist():
    profit_knapsack = profit_addition(profit_knapsack, p)

profit_greaterthan(profit_simple, profit_knapsack)

False

We can see that the output is False, which means using the 0/1 knapsack algorithm has a greater categorical profit than just picking the top 13 players.  

Discussion:

1. Weekly Player Performance Prediction: Time Series Predictions.  
    - I anticipate to have better results than using weighted average when using Time Series predictions, such as OLS Regression, Harmonic Regression, XGBoost, Arima, Vector Autoregression...  

2. Position constraint: 1PG, 1SG, 1G, 1SF, 1PF, 1F, 2C, 2Util, 3BN  
    -  This becomes a much more complex constraint optimization problem such as Mulitple Knapsack Assignment Problem (MKAP)

3. Snake Draft constraint: 
    - Simulate the snake draft: for simplicity, round1 only drafts within 1~12 picks, round2: 12~23, round3: 24~35,...   
    - Build a live draft version.  

4. Consider Accidents: Player Injuries, Game Time Decisions, Suspensions ...